### ISSUE TITLE
Too many open files

### ISSUE THREAD

**Issue by NuclearInstruments** [civetweb/civetweb#985]
Dear all,
I cross compiler the library on arm64 and works great except for a strange error.
I serve both web pages as static content and api with the library.
In the pages I have a JavaScript that interact with the API every seconds.
After a while, the server response with error 500 Too many opens file for the static content while perfectly works for the api (that do not opens file on disk)
If I check open sockets I see hundreds of open sockets associated to my app.
What could I investigate?

thank you
Andrea

**Comment by bel2125** [comment]
This seems strange, indeed 
You are not running out if file handles because you have open "files" but because of open "sockets"?

Probably you have `enable_keep_alive` set, but your client does only one request and does not close the socket on the other side either? Then the server keeps the socket open and waits for another request using the same connection. You could disable keep alive for a quick test.

Maybe the client does not close the connection properly? You can set some timeouts for this.

I don't know if you can build directly on the ARM device, to exclude cross build errors - although, I consider the reasons above as more likely.

**Comment by NuclearInstruments** [comment]
thank you
I investigated further more and it end up that,I'm running out due to the sockets. 

If I enter in console lsof i get tons of
4940	/mnt/fers_low_level_app.elf	socket:[3574973282]
24940	/mnt/fers_low_level_app.elf	socket:[3574973292]
24940	/mnt/fers_low_level_app.elf	socket:[3574973299]
24940	/mnt/fers_low_level_app.elf	socket:[3574973308]
24940	/mnt/fers_low_level_app.elf	socket:[3574973312]
24940	/mnt/fers_low_level_app.elf	socket:[3574973316]
24940	/mnt/fers_low_level_app.elf	socket:[3574973324]
24940	/mnt/fers_low_level_app.elf	socket:[3574973321]
24940	/mnt/fers_low_level_app.elf	socket:[3574973329]
24940	/mnt/fers_low_level_app.elf	socket:[3574964494]
24940	/mnt/fers_low_level_app.elf	socket:[3574973335]
24940	/mnt/fers_low_level_app.elf	socket:[3574973349]

I tried to set enable_keep_alive but it was not useful.

The problem is only present with AJAX call. If i load the API or static data in a infinite loop with wget the problem never appear.

these are my options. 
const char *options[] = { "document_root",
    						  _static_page_url.c_str(),
							  "listening_ports",
							  _port.c_str(),
							  "request_timeout_ms",
							  "10000",
							  "enable_keep_alive",
							  "no",
							  NULL};

ths function that close the connection if just this? mg_close_connection

Thank you
Andrea

**Comment by NuclearInstruments** [comment]
I added the following lines of code in oder to set a timeout of one second to to the closing sockets and for the moment seams to works.

	}
#endif
	if (conn->client.sock != INVALID_SOCKET) {


//Andrea Abba hack to set timeout to TIME_WAIT SOCKETs
struct linger sl;
sl.l_onoff = 1;		/* non-zero value enables linger option in kernel */
sl.l_linger = 0.1;	/* timeout interval in seconds */
setsockopt(conn->client.sock, SOL_SOCKET, SO_LINGER, &sl, sizeof(sl));
//end of Andrea Abba hack


#if defined(__ZEPHYR__)
		closesocket(conn->client.sock);
#else
		close_socket_gracefully(conn);
#endif
		conn->client.sock = INVALID_SOCKET;
	}

**Comment by bel2125** [comment]
`close_socket_gracefully()` already sets some linger timeout, if the `linger_timeout_ms` configuration is set - see [User Manual](https://github.com/civetweb/civetweb/blob/master/docs/UserManual.md#linger_timeout_ms). Did you try this?

**Comment by NuclearInstruments** [comment]
o ok i found it but in any case both my solution and use linger_timeout_ms just delays the problem. After less than an hour I got the same error. I'm investigating

**Comment by bel2125** [comment]
Your sockets are in the `TIME_WAIT` state (https://computingforgeeks.com/how-to-check-tcp-connections-states-in-linux-with-netstat/, http://tcpipguide.com/free/t_TCPOperationalOverviewandtheTCPFiniteStateMachineF-2.htm)?

A solution could be to close the connection from the server side.

Or to increase the number of allowed sockets. They will time out after 60 seconds on Linux (I think this is hard coded in some kernel versions, and can be modified by setting /proc/sys/net/ipv4/tcp_fin_timeout in others - but I'm not sure here, maybe give it a try).

**Comment by bel2125** [comment]
> A solution could be to close the connection from the server side.

Sorry, I mean client side - this could be a client issue as well, and easier to solve there:

http://tcpipguide.com/free/t_TCPConnectionTermination-1.htm
http://tcpipguide.com/free/t_TCPConnectionTermination-2.htm <--
http://tcpipguide.com/free/t_TCPConnectionTermination-3.htm
http://tcpipguide.com/free/t_TCPConnectionTermination-4.htm

You will still have to accept some sockets in a TIME_WAIT state at the server side. I don't know what your actual file descriptor limit is - in a somewhat protected environment (medium size intranet) I never had an issue with the Linux default setting of 1024 (what version of Linux?). If you have a much lower value, e.g. 16 file descriptors, you may run into trouble.

**Comment by stevebroshar** [comment]
I'm having similar if not same issue. Many sockets open and then app hits max file open limit (set to 1024) and then crashes. I have narrowed down the problem to HTTPS/SSL/TLS. Does not happen for non-encrypted. File handles are rock solid even with lots of clients when no encryption. But with encryption handles increase several a minute an the app crashes in about an hour.

I don't see how this would be anything but a bug in civet web. But, I'd be happy to change a setting to fix this.

Do you all still think this is just how it works? Or could it be a bug?

I see CIVETWEB_VERSION 1.10 in civetweb.h

**Comment by bel2125** [comment]
It could be also similar to this: https://github.com/civetweb/civetweb/issues/994

CivetWeb 1.10 is an older version (September 2017), could you use the current one? https://github.com/civetweb/civetweb/blob/master/RELEASE_NOTES.md#release-notes-v110
For V1.10 it could be an old bug that is already fixed.

If you have an issue with C or C++ handlers (your code): Could you try this approach: https://github.com/civetweb/civetweb/issues/994#issuecomment-827823448

**Comment by stevebroshar** [comment]
Thanks for the reply.

I am working to upgrade to 1.14 (which I think is the latest version). That is a complicated affair (at least it is for me); so it will take me a while...

As for the comment in #994: I understand the suggested change to use mg functions to format&send the header rather than formatting it in a more manual/custom way. But I don't understand how that will help. The comment says it will help with HTTP/2, and although that sounds nice, that is not my concern. The second part means little to me: "b) The "Connection: close" (or "Connection: keep-alive") header is sent automatically - and a "Date" header as well." It's sent automatically? Then, it's being sent? OK. So what? If it's being sent then what is there for me to do? Does it mean that _if_ I use these calls to format&send the header, _then_ it's sent automatically? Which call specifically results in the connection header being sent? How do I know whether it sends close or keep-alive?

**Comment by stevebroshar** [comment]
Upgrade not going well. When start server I get error: load_tls_dll: libssl.so: cannot find TLS_server_method, OPENSSL_init_ssl, TLS_client_method, SSL_CTX_set_options. I know this is about openSSL which is a huge PITA. Dependency hell. You (bell2125) are either a developer of civet web or a power user as I see your comments on https://github.com/civetweb/civetweb/issues/641. I wish I understood what you said. Seems important.

**Comment by stevebroshar** [comment]
And we use yocto which seems to limit the version we use. I'm told by co-worker that we get yocto bb from external source (don't write our own) and the newest version I can find is for civetweb 1.12. If that works, then we're all happy. If it has the same problem you're going to tell me to try the latest 1.14, but I don't know how to integrate that with our embedded build. :(

**Comment by bel2125** [comment]
CivetWeb itself can use either the OpenSSL 1.0.x API or the OpenSSL 1.1.x API - the version different seems minor, but in fact it's a major change and the two versions are incompatible (for programs using more than the most basic API functions). Since lots of developers seem to have their issues with OpenSSL, the most recent version (1.14) allows to use mbedTLS as an alternative - but this does not yet support all features OpenSSL would support.
The only dependency that really changed since V1.10 is the default setting what version of OpenSSL should be used. The old version defaults to OpenSSL 1.0.x, the new version defaults to 1.1.x (I think there is no further development on 1.0.x). You can still change this default by setting the OPENSSL_API_1_0, OPENSSL_API_1_1 (or OPENSSL_API_3_0 to use an OpenSSL 3.0 alpha version) - or USE_MBEDTLS.

You must set exactly one of these defines OPENSSL_API_1_0, OPENSSL_API_1_1, OPENSSL_API_3_0 or USE_MBEDTLS.
Otherwise you will get the default version, and that was OpenSSL 1.0.x up to CivetWeb 1.12 and OpenSSL 1.1.x starting from CivetWeb 1.13 (see [RELEASE_NOTES](https://github.com/civetweb/civetweb/blob/master/RELEASE_NOTES.md#release-notes-v113)). Nevertheless, you can still use 1.13 and 1.14 with OpenSSL 1.0.x (or 1.12 with OpenSSL 1.1.x), if you just set the appropriate define. Probably your external source (or coworker?) did not realize that setting this define will allow you to use a new version on an old operating system.
CivetWeb is often used for embedded systems that often do not have the most recent tools, APIs, ... so we are rather conservative with updates that will force you to use a new system (e.g., WinCE and WinXP support was dropped only recently). In other words, I think you should be able to run a new civetweb version on a Linux that did already run CivetWeb 1.12. And yes, I am one of the developers/maintainers of civetweb, and a power user as well, technology consultant ... and a couple of other things ;-) 
Regarding my "Connection: close" comment in the other thread: The idea is, if your server sends a "Connection: close" to the client, then the client will initiate the TCP shutdown sequence - it it is a "well behaved" client like a standard browser. This will cause the client to go into the "TIME WAIT" state, while the server does not need to wait for a timer. Just have a quick look at the state machine at the bottom of this site: http://www.tcpipguide.com/free/t_TCPOperationalOverviewandtheTCPFiniteStateMachineF-2.htm The close initiator will have to wait for a timer in the "TIME WAIT" state while the responder doe not. (If you are interested in more theory: http://www.serverframework.com/asynchronousevents/2011/01/time-wait-and-its-design-implications-for-protocols-and-scalable-servers.html)
 So, let the client be the initiator by adding a "Connection: close" to the response. You can either do this manually in any version of CivetWeb - if you use a new server and you are using the new API, the server will do it automatically. It will not avoid the TIME WAIT state in 100% of all cases, but you should see some improvement if you let the client close the connection.
The other tuning parameter is `linger_timeout_ms` --> see User Manual.

**Comment by stevebroshar** [comment]
I spent all week on this; much to the exasperation of my boss of the product guy. But, I have discovered that the problem in 1.10 (file handle leak) does not occur in 1.12 or 1.14. And, I think I've figured out how to build 1.14 (and 1.12) via yocto. So, things are coming together. I plan to use 1.14 unless I run into a hitch, but looks good so far. Using api 1.0. We can move to api 1.1 some other week. Maybe, I'll learn more about yocto so that such changes are not so time consuming. But, I keep telling people that if learn just a little more about yocto I'll know enough to be dangerous ... so best if don't learn more. I'm staying safe by staying ignorant :)

I will say that I tried to use the cmake stuff in the repo, but couldn't get that to work. The resulting lib didn't have the C++ entry points even though I selected the CPP feature. I am using the makefile instead. I don't know much about cmake. It seems overly complicated; just like yocto. The makefile seems well designed so I see no point making things more complicated. I read some about cmake and one fella said that cmake is for huge systems; saying something like no point tilling your herb garden with a massive farm implement.

I don't know what to say about the connection:close thing. I didn't know much about that before. I did some reading about it today and it seems that connection:keep-alive is desirable from performance perspective. We were using keep-alive so plan to leave that.

Thanks for the help bel2125.
